geniza/corpus/management/commands/merge_joins.py

```diff
@@ -0,0 +1,250 @@
+                for doc in documents:
+                    pgpid_match = self.re_see_pgpid.match(doc.description)
+                    if pgpid_match:
+                        print("*** pgpid match: %s" % pgpid_match.groupdict()["pgpid"])
+                        print(doc.id, doc.description)
+                        possible_primaries.add(int(pgpid_match.groupdict()["pgpid"]))
+
+                if possible_primaries:
+                    if len(possible_primaries) == 1:
+                        primary = list(possible_primaries)[0]
+                        if primary in group_ids:
+                            primary_id = primary
+                            status = "see PGPID"
+                            action = "MERGE"
+                        elif primary not in group_ids:
+                            print(shelfmark_type)
+                            print(documents)
+                            print("### primary id %s not in group" % primary)
+
+                    else:
+                        print(shelfmark_type)
+                        print(documents)
+                        print(
+                            "### more than one possible primary id: %s"
+                            % possible_primaries
+                        )
+
+                # also possible: see shelfmark
+
+            # other possibilities: if a subset of documents in a group match
+            # based on description, should they be merged?
+            for doc in documents:
+                doc_status = ""
+                if action == "MERGE":
+                    if doc.id == primary_id:
+                        doc_status = "primary"
+                    else:
+                        doc_status = "merge"
+
+                # keep track of status / action for report output
+                report_rows.append(
+                    [
+                        shelfmark_type,
+                        action,
+                        status,
+                        doc_status,
+                        doc.id,
+                        doc.description,
+                    ]
+                )
+
+        self.stdout.write("%d groups with the same description" % same_desc)
+
+        if self.mode == "report":
+            self.generate_report(report_rows)
+
+    def generate_report(self, report_rows):
+        # output report of what would be done when in report mode
+
+        with open("merge-report.csv", "w") as csvfile:
+            csvwriter = csv.writer(csvfile)
+            csvwriter.writerow(
+                [
+                    "merge group",
+                    "action",
+                    "status",
+                    "role",
+                    "pgpid",
+                    "description",
+                ]
+            )
+            csvwriter.writerows(report_rows)
+        self.stdout.write("Report of merge candidates output as merge-report.csv")
+
+    def merge_documents(primary_doc, merge_docs, rationale):
```

Comment:
I've been thinking about it, and I think `group_merge_candidates` should return a list of tuples of primary document, one or more merge documents, and text/string for the merge rationale. Then that list could be used either to generate the csv report or as input for actually doing the merge, depending on the mode. (The rationale would be used in the log entry documenting the change).

Reply:
Interesting...I guess that would allow you to manually remove false positive merges from the spreadsheet in between running the two commands, so that you only merged ones that you had reviewed. Makes sense!